File: FozHelpers.py
"""
Summary: General purpose utility functions (not project specific)
"""

import datetime
import feedparser
import grp
import httplib
import logging
import logging.handlers
import os
import pickle
import pwd
import re
import sys
import time
import traceback
import urllib
import urllib2

from urlparse import urlparse

class Struct(object):
    "Python doesn't have a struct class, so we will fake one"
    
    def __init__(self, **entries):
        self.__dict__.update(entries)
        
    def __repr__(self):
        args = ['%s=%s' % (k, repr(v)) for (k,v) in vars(self).items()]
        return '%s(%s)' % (self.__class__.__name__, ', '.join(args))

class Task(object):
    def __init__(self, name='', status='pending', percent=0, startTime='', endTime=''):
        self.name      = name
        self.status    = status
        self.percent   = percent
        self.startTime = startTime
        self.endTime   = endTime
        
    def __repr__(self):
        args = ['%s=%s' % (k, repr(v)) for (k,v) in vars(self).items()]
        return '%s(%s)' % (self.__class__.__name__, ', '.join(args))

    def __str__(self):
        return "%s\t%s\t%s\t%s\t%s" % (self.name, self.status, self.percent, self.startTime, self.endTime)
        
    def start(self):
        "Mark the current task as active"
        self.status = 'active'
        self.startTime = generateIso8601Date()
        
    def finish(self):
        "Mark the task as complete"
        self.status = 'complete'
        self.percent = 100
        self.endTime = generateIso8601Date()
        
    def fail(self):
        "Mark the task as failed"
        self.status = 'failed'
        self.endTime = generateIso8601Date()
        
    def clear(self):
        "Clear the and set it to pending"
        self.status    = 'pending'
        self.percent   = 0
        self.startTime = ''
        self.endTime   = ''
        

class Tasklist(object):
    def __init__(self, filename=None):
        self._tasks = []
        self._currentTask = -1
        self.filename = filename
        self.__status = 0
        
    def __iter__(self):
        for index in range(0, len(self._tasks), 1):
            yield self._tasks[index]
            
    def __repr__(self):
        args = ['%s=%s' % (k, repr(v)) for (k,v) in vars(self).items()]
        return '%s(%s)' % (self.__class__.__name__, ', '.join(args))
        
    def __str__(self):
        res = ""
        for task in self:
            res += str(task)+"\n"
        res=res[:-1]
        return res    
        
    def _getStatus(self):
        return self.__status
                
    def add(self, task):
        "Add a task object to the queue"
        
        argType = type(task).__name__
        
        if argType == 'str':
            self._tasks.append(Task(task))
        elif argType == 'Task':
            self._tasks.append(task)
        else:
            raise ValueError, "Can only add tasks to tasklist"
            
        return self
        
    def delete(self, taskNumber=None):
        "Delete a task from the queue (no task number passed in means delete the current task)"
        if taskNumber is None:
            taskNumber = self._currentTask
            
        if isBlank(self._tasks):
            raise ValueError, "Attempt to delete empty task"
            
        if taskNumber > len(self._tasks) - 1 or taskNumber < 0:
            raise ValueError, "attempt to delete invalid task"
            
        if taskNumber == self._currentTask:
            self.next()
            self._currentTask -= 1
            
        del(self._tasks[taskNumber])
        self.save()
        
        self.__status = int(float(self._currentTask) / len(self._tasks) * 100)
        
        return self
        
    def next(self):
        "Mark the current task as complete and advance the task counter"
        if self._currentTask >= 0:
            self._tasks[self._currentTask].finish()
            
        self._currentTask += 1
        
        if self._currentTask < len(self._tasks):
            self.__status = int(float(self._currentTask) / len(self._tasks) * 100)
            self._tasks[self._currentTask].start()
        else:
            self._currentTask = -1
            
        self.save()
            
    def finish(self):
        "Mark all remaining tasks complete and halt the queue"
        if self._currentTask == -1:
            return
            
        for index in range(self._currentTask, len(self._tasks), 1):
            self.next()
            
        self.__status = 100
        
        self.save()
            
    def abort(self):
        "Fail the current task and halt the queue by setting the task pointer to None"
        if self._currentTask < 0:
            self._currentTask = 0
            
        self._tasks[self._currentTask].fail()
        self._currentTask = -1
        self.save()
        
        return self
            
    def start(self):
        "Reset all the tasks in the queue to pending and move the task pointer to the beginning"
        # clear out the task list status flags and times on a start
        for index in range(0, len(self._tasks), 1):
            self._tasks[index].clear()
            
        self._currentTask = 0
        self._status = 0
        self._tasks[self._currentTask].start()
        self.save()
        
        return self
        
    def currentTask(self):
        "Return the current task object being worked on"
        if self._currentTask < 0:
            return None
        else:
            return self._tasks[self._currentTask]
            
    def save(self, filename=None):
        "Marshal ourselves to a file"
        if filename is None:
            filename = self.filename
            
        if filename is not None:
            contents = str(self._currentTask) + "\n"
            contents += str(self)
            fd = open(filename, 'w').write(contents)
            
            return self

    def load(self, filename=None):
        "Demarshal ourselves from a previously saved file"
        if filename is None:
            filename = self.filename
            
        if filename is not None:        
            fd = open(filename, 'r')
            vals = fd.readlines()
            fd.close()
            # now strip all the trailing newlines
            vals = map(lambda x: x.rstrip("\n"), vals)
            self.clear()
            # First line is the task pointer
            self._currentTask = int(vals.pop(0))
            # rest of vals contains the individual task
            for task in vals:
                (name, status, percent, startTime, endTime) = task.split("\t")
                self.add(Task(name=name, status=status, percent=percent, startTime=startTime, endTime=endTime))
            
    def clear(self):
        "Empty out the task list"
        self._tasks = []
        self._currentTask = -1
        
        return self
        
    status = property(lambda self: self._getStatus())
    
def setupLogging(logfile="./logfile.log", loglevel=logging.INFO, stdout=False, syslog=False, when='midnight', interval=1, backupCount=6):
    """
    Set up the logging service.  Returns a log object.  Set stdout and syslog to True to add add'l streams.
    Defaults to a timed rotating log file rotating daily at midnight with 6 backup copies.
    
    You can write to the log object with log.<level>("message").  Valid levels are (debug|info|warning|error|critical)
    """

    log = logging.getLogger(logfile)

    log.setLevel(loglevel)
    formatter = logging.Formatter('[%(asctime)-12s] %(levelname)-8s <%(module)s#%(funcName)s:%(lineno)s> %(message)s')

    # logging.handlers.RotatingFileHandler(filename, mode, maxBytes, backupCount, encoding) can be used for rotating based on file size
    filelog = logging.handlers.TimedRotatingFileHandler(logfile, when=when, interval=interval, backupCount=backupCount)
    filelog.setFormatter(formatter)

    log.addHandler(filelog)
    
    if syslog:
        sl = logging.handlers.SysLogHandler(('localhost',514))
        sl.setFormatter(formatter)
        log.addhandler(sl)
        
    if stdout:
        so = logging.StreamHandler(sys.stdout)
        so.setFormatter(formatter)
        log.addHandler(so)

    return log

def isBlank(arg):
    argType = type(arg).__name__
    
    
    if argType == 'str':
        return True if len(arg.strip()) == 0 else False
    elif argType == 'list' or argType == 'dict':
        return True if len(arg) == 0 else False
    elif argType == 'NoneType':
        return True
    else:
        raise ValueError, "illegal type: %s" % (argType)
        
def fetchRemoteFile(url, localFile=None, permissions=None, owner=None, group=None, retries=0):
    "Fetch a remote file and either write it to a local file or return it"
    # if token:
    #     cookie = "iPlanetDirectoryPro=%s" % (token)
    # else:
    #     cookie = None
    while retries > -1:
        try:
            if localFile:
                contents = urllib.URLopener()
                # if cookie:
                #     contents.addheader('Cookie', cookie)
                contents.retrieve(url, localFile)
                contents.close()
                if permissions:
                    os.chmod(localFile, permissions)
                if owner or group:
                    chown(localFile, owner=owner, group=group)
                return
            else:
                request = urllib2.Request(url)
                if cookie:
                    request.add_header('Cookie', cookie)
                response = urllib2.urlopen(request)
                fileContents = response.read()
                actualUrl = response.geturl()
                if actualUrl != url:
                    raise IOError, ('http error',302, "resource was redirected")
                response.close()
                return fileContents
        except Exception, e:
            # Ignore exceptions, we will throw an exception if retries
            # are exceeded only
            if retries != 0:
                retries -= 1
                pass
            else:
                raise

def formatExcPlus(exc_info):
    """
    Return descriptive exception info, including all locals, frame info and traceback.
    Pass this method an exc_info object and it will pretty print an entire stack trace
    with values of every local variable in scope at each stack level.
    """    
    (type_, value_, tb) = exc_info
    
    tb_info = "Exception type: %s\nValue: %s\n\nTraceback Info:\n" % (type_, value_)
    
    tb_info += "".join(traceback.format_tb(tb))
    tb_info += "\n"
    
    while True:
        if not tb.tb_next:
            break
        tb = tb.tb_next
    stack = []
    f = tb.tb_frame
    while f:
        stack.append(f)
        f = f.f_back
    stack.reverse()

    tb_info += "Locals by frame, innermost last\n"
    for frame in stack:
        tb_info += "\n"
        tb_info += "Frame %s in %s at line %s\n" % (frame.f_code.co_name,
            frame.f_code.co_filename, frame.f_lineno)
        for key, value in frame.f_locals.items():
            tb_info += "\t%20s = " % (key)
            try:
                tb_info += "%s\n" % (str(value))
            except:
                tb_info += "<ERROR WHILE PRINTING VALUE>\n"
    return tb_info
                
def grep(string,list,opt=''):
    "search a list for a substring.  Pass -v as an option to return everything BUT the match"
    
    expr = re.compile(string)
    if opt=='-v':
        nexpr = lambda x: expr.search(x) == None
        return filter(nexpr,list)
    return filter(expr.search,list)

def timestamp():
    return datetime.datetime.today().strftime("%D %T")
        
def generateDate(offsetDays=0):
    "alias function for generateHttpDate"
    return generateHttpDate(offsetDays)
    
def generateHttpDate(offsetDays=0):
    "Generate an httpdate formatted date string, offsetDays from now (in UTC/GMT)."
    return (datetime.datetime.utcnow() + datetime.timedelta(days=offsetDays)).strftime('%a, %d %b %Y %T GMT')

def generateIso8601Date(offsetDays=0):
    "Generate an ISO-8601 formatted date string in UTC, offsetDays from now yyyy-mm-dd'T'HH:MM:SS.fff'Z'."
    
    return (datetime.datetime.utcnow() + datetime.timedelta(days=offsetDays)).isoformat()+"Z"
    
def generateInternetTime():
    "Generate a timestamp in internet time format (beats)"
    t =  datetime.datetime.utcnow() + datetime.timedelta(seconds=3600) # Biel, Switzerland
    midnight = datetime.datetime(year=t.year, month=t.month, day=t.day)
    secs = (t - midnight).seconds
    beats = int(secs/86.4)
    return beats

def writeFile(filename, contents='', owner=None, group=None, permissions=None, append=False):
    # os.chmod() second param is OCTAL
    # python automatically converts 0644 from octal into decimal
    
    close = False
    
    if type(filename).__name__ == "str":
        # we got passed a filename to create
        if append:
            mode = "a"
        else:
            mode = "w"
            
        fd = open(filename, mode)
        name = fd.name
        # if we create the file, we close it, otherwise we leave it open
        close = True
    elif type(filename).__name__ == "file":
        # we got passed an open file descriptor
        fd = filename
        name = fd.name
    elif type(filename).__name__ == 'instance' and type(filename.file).__name__ == 'file':
        # we got passed a temporary file instance created by the tempfile module
        # which isn't 'quite' the same as a normal FD although it acts like one
        fd = filename
        name = filename.name
    else:
        raise TypeError, "filename must be a file object or a filename (string)"
        
    fd.write(contents)
    
    if close:
        fd.close()

    if permissions:
        os.chmod(name, permissions)
        
    if owner or group:
        chown(name, owner, group)

def chown(filename, owner=None, group=None):
    # -1 to chown means don't change it
    
    uid = pwd.getpwnam(owner).pw_uid if owner else -1
    gid = grp.getgrnam(group).gr_gid if group else -1

    os.chown(filename, uid, gid) 

def getEc2PublicHostName():
    return fetchRemoteFile("http://169.254.169.254/2009-04-04/meta-data/public-hostname")
    
#####
##### SSO Methods
#####

def getSSOToken(uri, username, password):
    """
    Attempt to retrieve a valid token from an SSO server.
    Pass in the base URI to the sso realm (e.g. http://ssoserver.foo.com/opensso)
    and the username and password.  Returns None if invalid login
    """
    
    uri += "/identity/authenticate?username=%s&password=%s" % (username, password)
    
    try:
        req = urllib2.Request(uri)
        response = urllib2.urlopen(req)
        body = response.read()
        token = body.split('=', 1)[1].rstrip("\n")
    except urllib2.HTTPError, e:
        if e.code == 401:
            # Invalid login
            return None
        else:
            raise
    else:
        return token

def verifySSOToken(uri, token):
    "Attempt to verify a token against a base SSO uri"
    
    uri += "/identity/isTokenValid"

    data = urllib.urlencode({'tokenid': token})

    try:
        req = urllib2.Request(uri, data)
        response = urllib2.urlopen(req)
        page = response.read()
    except urllib2.HTTPError, e:
        if e.code == 401:
            valid=False
        else:
            raise
    else:
        body = page.split("=", 1)[1].rstrip("\n")
        if body.lower() == 'true':
            valid = True
        else:
            valid = False

    return valid
    
def invalidateSSOToken(uri, token):
    "Logout and invalidate a previously generated token"
    
    uri += "/identity/logout"

    data = urllib.urlencode({'subjectid': token})

    try:
        req = urllib2.Request(uri, data)
        response = urllib2.urlopen(req)
    except urllib2.HTTPError, e:
        return e.code
    else:
        return response.code

def authorizeSSOResource(uri, token, action):
    "Determine if given action is authorized to token at uri"

    uri += "/identity/authorize"

    if action.upper() in ['GET', 'PUT', 'POST', 'DELETE', 'OPTIONS', 'HEAD']:
        try:
            data = urllib.urlencode({'uri': uri, 'action': action.upper(), 'subjectid': token})
            req = urllib2.Request(uri, data)
            response = urllib2.urlopen(req)
            page = response.read()
        except urllib2.HTTPError, e:
            valid = False
        else:
            body = page.split("=", 1)[1].rstrip("\n")
            if body.lower() == 'true':
                valid = True
            else:
                valid = False
    else:
        raise ValueError, "Action must be one of GET/POST/PUT/DELETE/OPTIONS/HEAD"

    return valid
    
def urlEscape(uri):
    "Properly escapes a URI string and returns the escaped version"
    
    return urllib.quote(urllib.unquote(uri), safe=":/")

def installPackage(packages):
    "accepts either a list or a single package and installs it"
    
    if type(packages).__name__ == 'str':
        packages = [packages]
        
    for pkg in packages:
        rc = runCmd("! /usr/bin/yum install -y --nogpgcheck %s | egrep '^No package .+ available\.' 2>&1 >/dev/null " % (pkg))
        if rc:
            # if we get an error, exit prematurely
            return rc
            
def writeTaskProgress(filename, caller, taskName, percentComplete, startTime=generateIso8601Date()):
    fd = open(filename, "a")
    fd.write("%s\t%s\t%s\t%s\t%s\n" % (generateIso8601Date(), caller, percentComplete, startTime, taskName))
    fd.close()

def getProgressLogName():
    me = os.path.basename(sys.argv[0])
    me = "/opt/nodeagent/handlers/state/" + re.sub(r'.template.*$', '', me) + ".status"
    return me
    
def exitIf(rc):
    if rc != 0:
        sys.exit(rc)

def runCmd(command, ignoreRC=True, asUser=None):
    if asUser:
        # Need to escape any double quotes
        if command.find('"') != -1:
            command = command.replace('"', '\\"')
        command = "/sbin/runuser -s /bin/bash - %s -c \"%s\"" % (asUser, command)
        
    rc = os.system(command)
    if not ignoreRC and rc:
        raise OSError, "%s returned a non-zero RC: %d" % (command, rc)
    else:
        return rc
        
def addUser(username, uid=None, group=None, shell=None, homeDir=None, createHome=False):
    cmd = "/usr/sbin/useradd "
    cmd += "-u %d" % (uid) if uid else " "
    cmd += "-g %s" % (group) if group else " "
    cmd += "-s %s " % (shell) if shell else " "
    cmd += "-d %s " % (homeDir) if homeDir else " "
    cmd += "-m " if createHome else "-M "
    cmd += username
    
    rc = runCmd(cmd)
    return rc

def addGroup(groupname, gid=None):
    cmd = "/usr/sbin/groupadd "
    cmd += "-g %d" % (gid) if gid else " "
    cmd += groupname
    
    rc = runCmd(cmd)
    return rc

def setPassword(username, password):
    cmd = "echo '%s' | /usr/bin/passwd --stdin %s" % (password, username)
    rc = runCmd(cmd)
    return rc
    
def removeTimestamp(filename):
    # Remove timestamp added by webresource , e.g 2011-01-25T13-31-20-latest.tar.gz
    import re
    tsPattern=r'^(19|20)\d\d[- / .](0[1-9]|1[012])[- / .](0[1-9]|[12][0-9]|3[01])T(0[0-9]|1[0-9]|2[0-4])[- / .](0[0-9]|1[0-9]|2[0-9]|3[0-9]|4[0-9]|5[0-9]|60)[- / .](0[0-9]|1[0-9]|2[0-9]|3[0-9]|4[0-9]|5[0-9]|60)-'
    
    return re.sub(tsPattern, '', filename)
    

File: FozHelpers.rb
# This is a collection of useful hacks, monkeypatches, class extensions, overrides, etc.

require 'rubygems'
require 'enumerator'
require 'rss'
require 'open-uri'
require 'net/http'
require 'etc'
require 'fileutils'
require 'term/ansicolor'

####
#### Useful constants
####

EMAIL_REGEXP =  Regexp.new(/[^ @]+@[^ @]+\.[^ @]+/)

#####
##### Generic monkeypatches to extend existing base classes
#####

class String
   # Give us the ability to do things like puts 'foo'.red.bold
   include Term::ANSIColor

   def shift
      return nil if self.empty?
      item=self[0]
      self.sub!(/^./,"")
      return nil if item.nil?
      item.chr
   end

   def unshift(other)
      newself = other.to_s.dup.pop.to_s + self
      self.replace(newself)
   end

   def pop
      return nil if self.empty?
      item=self[-1]
      self.chop!
      return nil if item.nil?
      item.chr
   end

   def push(other)
      newself = self + other.to_s.dup.shift.to_s
      self.replace(newself)
   end

   def rotate_left(n=1)
      n=1 unless n.kind_of? Integer
      n.times do
         char = self.shift
         self.push(char)
      end
      self
   end

   def rotate_right(n=1)
      n=1 unless n.kind_of? Integer
      n.times do
         char = self.pop
         self.unshift(char)
      end
      self
   end

   @@first_word_re = /^(\w+\W*)/
   @@last_word_re  = /(\w+\W*)$/

   def shift_word
      # shifts first word off of self
      # and returns; changes self
      return nil if self.empty?
      self=~@@first_word_re
      newself = $' || ""   # $' is POSTMATCH
      self.replace(newself) unless $`.nil?
      $1.strip
   end

   def acronym(thresh=0)
      self.split.find_all {|w| w.length > thresh }.
      collect {|w| w[0,1].upcase}.join
   end

   def unshift_word(other)
      # Adds provided string to front of self
      newself = other.to_s + " " + self
      self.replace(newself)
   end

   def pop_word
      # pops and returns last word off self
      # changes self
      return nil if self.empty?
      self=~@@last_word_re
      newself=$` || ""  # $` is PREMATCH
      newself.rstrip!
      self.replace(newself) unless $`.nil?
      $1
   end

   def push_word(other)
      # pushes provided string onto end of self
      newself = self + " " + other.to_s
      self.replace(newself)
   end

   def rotate_word_left
      word = self.shift_word
      self.push_word(word)
   end

   def rotate_word_right
      word = self.pop_word
      self.unshift_word(word)
   end

   # Patch the string class to give a rightstr function
   # Call with String.rightstr(length)

   def rightstr(length)
      if length < 0
         raise ArgumentError, "length cannot be a negative value"
      end
      self.reverse[0,length].reverse
   end

   # Create an each_char iterator that will return 'n' characters each yield

   def each_char(n=1)
      if n < 1
         raise ArgumentError, "length must be > 0"
      end

      if block_given?
         split(//).each_slice(n) do |x|
            yield x.join
         end
      else
         results = Array.new
         split(//).each_slice(n) {|x| results.push(x.join)}
         results
      end
   end

   def blank?
      !(self =~ /\S/)
   end

   # add a coerce function to the String class so we can get automatic type conversion for arithmetic
   # which makes stuff like 1 + "2" work without having to cast
   def coerce(other)
      case other
      when Integer
         begin
            return other, Integer(self)
         rescue
            return Float(other), Float(self)
         end
      when Float
         return other, Float(self)
      else
         super
      end
   end
end

class Fixnum
   def odd?
      (self % 2 == 1) ? true : false
   end
end

class Array
   # Snag a random element from an array
   def rand
      self[Kernel::rand(self.size)]
   end

   # Shuffle the array
   def randomize
      sort {|a,b| Kernel::rand <=> 0.5}
   end

   def randomize!
      replace randomize
   end
end

class Numeric
   # Time calculations
   #   i.e. Numeric.minutes, etc
   # These are ripped out of ActiveSupport and give us things like 2.hours.since(4.hours.ago) and 5.minutes.from_now

   def seconds
      self
   end

   def minutes
      self * 60
   end
   alias :minute :minutes

   def hours
      self * 60.minutes
   end
   alias :hour :hours

   def days
      self * 24.hours
   end
   alias :day :days

   def weeks
      self * 7.days
   end
   alias :week :weeks

   def ago(time = Time.now)
      time - self
   end
   alias :until :ago

   def since(time = Time.now)
      time + self
   end
   alias :from_now :since
end

class Time
   # Monkeypatch the time class to include other conversions

   def internet_time
      t = self.gmtime + 3600 # Biel, Switzerland
      midnight = Time.gm(t.year, t.month, t.day)
      secs = t - midnight
      beats = (secs/86.4).to_i
   end

   def timestamp
      self.strftime("%D %T")
   end

   def iso8601
      # Generate an ISO-8601 formatted date string in UTC, offsetDays from now yyyy-mm-dd'T'HH:MM:SS.fff'Z'.
      self.w3cdtf[/^.+\.\d+/] + "Z"
   end
end

#####
##### Various helpful classes
#####


### Basic node and linked list class
### LL class implements push, pop, and iterators
### as well as convenience getters for LinkedList#last and #first

class Node
   attr_reader :value, :ptr

   def initialize(value, ptr)
      @value, @ptr = value, ptr
   end
end

class LinkedList
   include Enumerable

   attr_reader :size, :head

   def initialize
      clear
      self
   end

   def clear
      @head, @tail, @size = nil, nil, 0
   end

   def empty?
      head.nil?
   end

   def push(value)
      @head = Node.new(value, head)
      @size += 1
      @last = @head if @last.nil?
      self
   end

   def pop
      return nil if empty?
      @size -= 1
      value = head.value
      @head = head.ptr
      @last = nil if empty?
      value
   end

   def append(value)
      replace reverse.push(value).reverse
   end

   def reverse
      new_list = self.class.new
      self.each { |x| new_list << x }
      new_list
   end

   def reverse!
      replace reverse
   end

   def each
      return nil if empty?

      node = head
      loop do
         yield node.value
         node = node.ptr
         break unless node
      end
   end

   def last
      return nil if empty?
      @last.value
   end

   def first
      return nil if empty?
      @head.value
   end

   def ==(other)
     head == other.head
   end

   def replace(other)
      new_list = self.dup
      clear
      other.reverse.each { |x| push(x) }
      self
   end

   alias_method :<<, :push
   alias_method :unshift, :push
   alias_method :length, :size
   alias_method :shift, :pop
end

class Task < Object

   attr_accessor :name, :status, :percent, :start_time, :end_time

   def initialize(name=nil, *options)
      if options.empty?
         options = {}
      else
         options = options[0]
      end

      @name        = name

      @status      = options[:status]     || 'pending'
      @percent     = options[:percent]    || 0
      @start_time  = options[:start_time] || ''
      @end_time    = options[:end_time]   || ''
   end

   def to_s
      "#{@name}\t#{@status}\t#{@percent}\t#{@start_time}\t#{@end_time}"
   end

   def start
      # Mark the current task as active"
      @status     = 'active'
      @start_time = Time.now.iso8601
      @end_time   = ''
      @percent    = 0
      self
   end

    def finish
       # Mark the task as complete
       @status   = 'complete'
       @percent  = 100
       now = Time.now.iso8601
       if @start_time.blank?
           @start_time = now
       end
       @end_time = now
       return self
    end

   def fail
      # Mark the task as failed
      @status   = 'failed'
      @end_time = Time.now.iso8601
      self
   end

   def clear
      # Clear the task and set it to pending
      @status     = 'pending'
      @percent    = 0
      @start_time = ''
      @end_time   = ''
      self
   end
end

class Tasklist
   protected

   attr_writer :current_task

   public

   attr_accessor :filename

   # getters

   def current_task
      @tasks[@current_task]
   end

   def status
      @status
   end

   # constructors, iterators, etc.

   def initialize(filename=nil)
        @tasks = []
        @current_task = -1
        @filename = filename
        @status = 0
        self
   end

   def each
      @tasks.each { |task| yield task }
   end

   def each_with_index
      @tasks.each_with_index {|task, index| yield [task, index] }
   end

   alias :with_index :each_with_index

   def to_s
      @tasks * "\n"
   end

   # class methods

   def add(task)
      if task.instance_of? String
         t = Task.new(task)
      elsif task.instance_of? Task
         t = task
      else
         raise TypeError, "Can only add tasks to tasklist"
      end

      @tasks << t

      self
   end

   def delete(task_number=@current_task)
      if @tasks.empty?
         # Should throw an exception here
         raise ArgumentError, "attempt to delete from an empty tasklist"
      end

      if task_number > @tasks.length - 1 or task_number < 0
         # throw an exception
         raise ArgumentError, "attempt to delete an invalid task"
      end

      if task_number == @current_task
         self.next
         @current_task -= 1
      end

      @tasks.delete_at(task_number)
      self.save

      @status = (@current_task + 1.0 / @tasks.length * 100).to_i

      self
   end

   def next
      if @current_task != -1
         @tasks[@current_task].finish
         @status = (@current_task + 1.0 / @tasks.length * 100).to_i
      end

      @current_task += 1

      if @current_task < @tasks.length
         @tasks[@current_task].start
      else
         @current_task = -1
      end

      self.save
      self
   end

   def abort
      # mark the current task as failed and reset the task pointer
      if @current_task == -1
         self
      end

      @tasks[@current_task].fail
      @current_task = -1
      self.save
      self
   end

   def finish
      # mark all outstanding jobs as finished by incrementing through the task list
      true while self.next

      self.save
      self
   end

   def start
      # Reset all the tasks in the queue to pending and move the task pointer to the beginning

      @tasks.collect! {|x| x.clear}

      @current_task = 0
      @status = 0
      @tasks[@current_task].start

      self.save
      self
   end

   def clear
      @tasks = []
      @current_task = -1
      self
   end

   def save(filename=nil)
      # Persist ourself to a filename

      filename = @filename if filename.nil?

      if not filename.nil?
         marshal = @current_task.to_s + "\n"
         marshal << self.to_s

         File.open(filename, "w") do |fd|
            fd.write(marshal)
         end
      end
      self
   end

   def load(filename=nil)
      filename = @filename if filename.nil?

      if not filename.nil?
         rep = self.class.new
         File.open(filename, "r") do |fd|
            rep.current_task = fd.gets.to_i
            while line = fd.gets
               line.chomp!
               (name, status, percent, start_time, end_time) = line.split("\t")
               rep.add(Task.new(name, :status=>status, :percent=>percent, :start_time=>start_time, :end_time=>end_time))
            end
         end
         rep
      end
   end

   def load!(filename=nil)
      filename = @filename if filename.nil?

      if not filename.nil?
         File.open(filename, "r") do |fd|
            clear
            @current_task = fd.gets.to_i
            while line = fd.gets
               line.chomp!
               (name, status, percent, start_time, end_time) = line.split("\t")
               self.add(Task.new(name, :status=>status, :percent=>percent, :start_time=>start_time, :end_time=>end_time))
            end
         end
         self
      end
   end
end

#####
##### Generic (non class) methods
#####

module Helpers
   def Helpers.fetch_remote_file(url, *options)
      if options.empty?
         options = {}
      else
         options = options[0]
      end

      local_file   = options[:local_file] || nil
      permissions = options[:permissions] || nil
      owner       = options[:owner] || nil
      group       = options[:group] || nil

      # Fetch a remote file and either write it to a local file or return it

      response = URI::parse(url)
      ifd = response.open()

      if local_file.nil?
         filecontents = ifd.read()
         ifd.close()
         filecontents
      else
         ofd = open(local_file, "w")
         bytes = File::copy_stream(ifd, ofd)
         ifd.close()
         ofd.chmod(permissions)
         ofd.close()
         Helpers.chown(local_file, owner, group)
      end
      bytes
   end

   def Helpers.get_ec2_public_hostname
      Helpers.fetch_remote_file("http://169.254.169.254/2009-04-04/meta-data/public-hostname")
   end

   def Helpers.write_file(filename, contents, *options)
      if options.empty?
         options = {}
      else
         options = options[0]
      end

      owner       = options[:owner] || nil
      group       = options[:group] || nil
      permissions = options[:permissions] || nil

      close = false

      if filename.instance_of?(String)
         # we got passed a filename to create
         fd = File.new(filename, "w")
         # if we create the file, we close it, otherwise we leave it open
         close = true
      elsif filename.instance_of?(File)
         # we got passed an open file descriptor
         fd = filename
      else
         raise TypeError, "filename must be a file object or a filename (string)"
      end

      fd << contents

      if permissions
         fd.chmod(permissions)
      end

      unless owner.nil? and group.nil?
         Helpers.chown(fd, owner, group)
      end

      if close
         fd.close()
      end
      return true
   end

   def Helpers.chown(filename, owner, group)
      uid = Etc.getpwnam(owner).uid unless owner.nil?
      gid = Etc.getgrnam(group).gid unless group.nil?

      if filename.instance_of? String
         FileUtils.chown(uid, gid, filename)
      elsif filename.instance_of? File
         filename.chown(uid, gid)
      else
         raise TypeError, "invalid filetype (must be open fd or string)"
      end
      return true
   end

   def Helpers.run_cmd(cmd)
      results = %x[cmd]
      $?
   end

   def Helpers.get_progress_log_name
      me = File::basename($0)[/^.*?(?=\.template)/]
      "/opt/nodeagent/handlers/state/#{me}.status"
   end

   def Helpers.add_user(username, *options)
      if options.empty?
         options = {}
      else
         options = options[0]
      end

      shell       = options[:shell] || nil
      home        = options[:home] || nil
      create_home = options[:create_home] || false

      cmd = "useradd "
      cmd += "-s #{shell} " unless shell.nil?
      cmd += "-d #{homeDir} " unless home.nil?
      cmd += (create_home) ? "-m " : "-M "
      cmd += username

      Helpers.run_cmd(cmd)
   end

   def Helpers.install_packages(packages)
      if packages.instance_of? String
         packages = [packages]

      for pkg in packages
          rc = Helpers.run_cmd("! /usr/bin/yum install -y --nogpgcheck %s | egrep '^No package .+ available\.' 2>&1 >/dev/null " % (pkg))
          if rc
             return rc
          end
       end
       true
    end
   end
end
